text/0022-remotes.md

```diff
@@ -0,0 +1,98 @@
+
+Currently functionality that exists only in the Registry now includes:
+
+* `POST /dataset` to upload a dataset's head
+* `/search` to search over known datasets (online or not)
+* `/datasets` to list all known datasets (online or not)
+* `/profile` user information
+* `/reputation` user trustability
+* `/pins` pinning a dataset
+* `/dsync` data block uploading
+
+Long-term we want to obliviate the need for these to be in a separate executable.
+
+The plan of action is as follows:
+
+* Add a flag to Qri's backend to run it as a remote. This will work similarily to the current "read-only" flag, disabling most write functionality.
+* Add APIs to the backend that search a similar purpose to what the Registry is doing now: `POST /dataset` in particular.
+* Get enough APIs in the backend such that it can be run as a remote which provides availability to users that want it.
+* Add functionality to Qri's command-line to publish to this remote instead of, or in addition to the public Registry.
+* Over time, move the rest of the functionality from the Registry into the backend remote.
+* Replace the live Registry with a standard remote.
+* Delete old Registry code.
+* New semi-centralized features get added to the Remote feature set.
+
+Some specific notes as to how this is presented. First, the concept of the Registry is a good way to explain Qri to users: it exists as a way to keep data online even if your own node is not. This shouldn't change as we move to Remotes; instead of deleting the _concept_ of a Registry, we're generalizing it by introduced a Remote as something you can run yourself if you're an advanced user with a special use-case. Secondly, since a Remote is considered an expert-level concept, it's okay for its UX to be slightly more advanced than it may be with other features. The easy thing to use should remain to be the Qri run public Registry. This implies, for example, that setting up a client to use a remote should be more like `git config set remote <url>` instead of `git remote add <name> <url>`, with `remote` as a top-level command.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Posting a dataset to a Remote involves expanding our current protocol:
+
+* Client sends Dataset Manifest and Head to the Remote, signed with their cryptographic key
+* The Remote has some ruleset about what it accepts, for example, no more than 10Mb in a body
+* Remote replies with what it wants from the user
+* DAGSync occurs (or not, if the dataset body is rejected)
+* Dataset is pinned
+* Remote tells the client about the result
+
+In the future, advanced users may want to configure a Remote with full granularity how an acceptance ruleset works. Perhaps they could even use `starlark` as a way to decide this.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Centralization is bad, which is why we need to be careful to make Remotes remain an optional feature that users are not forced into.
```

Comment:
Don't remotes create a federated model rather than a centralized model, though? Is Qri strongly opinionated that only p2p is a good idea here?

Reply:
Agreed, this remotes RFC creates something that looks a lot like a federated model, and is also less centralized than the way Qri currently works. This RFC is moving more privileged capabilities onto peers.

The only thing that beats a "pure decentralized" approach for us is the requirement that it work consistently. So no, we're not "strong p2p" here, but will head in that direction whenever we can 😄. We'll add some clarifying details here.